### PR TITLE
feat(metadata): Create metadata updating V2 DTOs

### LINKED ISSUE #258: Create metadata DTOs and data models for V2 Updating API
Because the PR https://github.com/edgexfoundry/go-mod-core-contracts/issues/249 is too big and updating API is more complicated, so create this issue for updating API.

This issue should create DTOs for Updating API:
* Device
* DeviceProfile
* DeviceService

### PR DESCRIPTION
**Pull request by weichou1229**
Fix #258 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #258 

## What is the new behavior?
Create device and deviceService, device profile updating DTOs for v2 API.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

File: v2/dtos/requests/deviceservice.go [hunk 1]
```diff
@@ -60,3 +62,121 @@ func AddDeviceServiceReqToDeviceServiceModels(addRequests []AddDeviceServiceRequ
...
+
+	// validate AddDeviceServiceRequest DTO
+	if err := ds.Validate(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ReplaceDeviceServiceModelFieldsWithDTO replace model's fields with DTO patch
+func ReplaceDeviceServiceModelFieldsWithDTO(ds *models.DeviceService, patch dtos.UpdateDeviceServiceDTO) {
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceservice.go:100]: So idea is the DB update will load existing `DeviceService` model from DB, call this to update the fields that changed and then store the update  `DeviceService` model to DB?

File: v2/dtos/requests/deviceservice.go [hunk 2]
```diff
@@ -60,3 +62,121 @@ func AddDeviceServiceReqToDeviceServiceModels(addRequests []AddDeviceServiceRequ
...
+	//	}
+	//}
+	//if baseAddress, exist := patch["baseAddress"]; exist {
+	//	v, ok := baseAddress.(string)
+	//	if ok {
+	//		ds.BaseAddress = v
+	//	}
+	//}
+
+	vals := reflect.ValueOf(ds)
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceservice.go:161]: I agree the the other approach with pointers is cleaner and much easier to follow.

File: v2/dtos/requests/deviceservice.go (hunk 1 above)
Comment (weichou1229) [reply v2/dtos/requests/deviceservice.go:100]: Yes, I thought the metadata handler will invoke dbClient to load the existing DeviceService and replace the field value with patch, then store into the DB.

File: v2/dtos/requests/deviceservice.go (hunk 2 above)
Comment (weichou1229) [reply v2/dtos/requests/deviceservice.go:161]: Thanks, I will use the pointer in implementation and update the swagger file with the issue https://github.com/edgexfoundry/edgex-go/issues/2639.

File: v2/dtos/requests/deviceprofile_test.go [hunk 3]
```diff
@@ -225,9 +249,176 @@ func TestAddDeviceProfile_UnmarshalYAML(t *testing.T) {
...
+	fmt.Println(string(resultTestBytes))
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var expected = tt.addDeviceProfile
+			err := tt.addDeviceProfile.UnmarshalJSON(tt.args.data)
+			if tt.wantErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, expected, tt.addDeviceProfile, "Unmarshal did not result in expected AddDeviceProfileRequest.", err)
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceprofile_test.go:284]: Change `AddDeviceProfileRequest` => `UpdateDeviceProfileRequest`

File: v2/dtos/requests/deviceprofile_test.go [hunk 4]
```diff
@@ -225,9 +249,176 @@ func TestAddDeviceProfile_UnmarshalYAML(t *testing.T) {
...
+        "requestId":"7a1707f0-166f-4c4b-bc9d-1d54c74e0137",
+		"profile":{
+			"name":"test device profile"
+		}
+	}`
+	var req UpdateDeviceProfileRequest
+
+	err := req.UnmarshalJSON([]byte(reqJson))
+
+	assert.Equal(t, nil, err)
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceprofile_test.go:376]: use `require.No Error(...)`

File: v2/dtos/requests/deviceprofile_test.go [hunk 5]
```diff
@@ -225,9 +249,176 @@ func TestAddDeviceProfile_UnmarshalYAML(t *testing.T) {
...
+	// Nil field checking is used to update with patch
+	assert.True(t, req.Profile.Manufacturer == nil)
+	assert.True(t, req.Profile.Description == nil)
+	assert.True(t, req.Profile.Model == nil)
+	assert.True(t, req.Profile.Labels == nil)
+	assert.True(t, req.Profile.DeviceResources == nil)
+	assert.True(t, req.Profile.DeviceCommands == nil)
+	assert.True(t, req.Profile.CoreCommands == nil)
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceprofile_test.go:384]: Use `assert.Nil(...)` for these

File: v2/dtos/requests/deviceprofile_test.go [hunk 6]
```diff
@@ -225,9 +249,176 @@ func TestAddDeviceProfile_UnmarshalYAML(t *testing.T) {
...
+	}`
+	var req UpdateDeviceProfileRequest
+
+	err := req.UnmarshalJSON([]byte(reqJson))
+
+	assert.Equal(t, nil, err)
+	// Empty slice is used to remove the data
+	assert.True(t, req.Profile.Labels != nil)
+	assert.True(t, req.Profile.DeviceCommands != nil)
+	assert.True(t, req.Profile.CoreCommands != nil)
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceprofile_test.go:405]: Use `assert.NotNil(...)` for these

File: v2/dtos/requests/deviceprofile_test.go [hunk 7]
```diff
@@ -225,9 +249,176 @@ func TestAddDeviceProfile_UnmarshalYAML(t *testing.T) {
...
+			"labels":[],
+			"deviceCommands":[],
+			"coreCommands":[]
+		}
+	}`
+	var req UpdateDeviceProfileRequest
+
+	err := req.UnmarshalJSON([]byte(reqJson))
+
+	assert.Equal(t, nil, err)
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceprofile_test.go:401]: use `require.NoError(....)`

File: v2/dtos/requests/deviceservice_test.go [hunk 8]
```diff
@@ -115,3 +138,122 @@ func Test_AddDeviceServiceReqToDeviceServiceModels(t *testing.T) {
...
+	}
+	tests := []struct {
+		name             string
+		addDeviceService UpdateDeviceServiceRequest
+		args             args
+		wantErr          bool
+	}{
+		{"unmarshal AddDeviceServiceRequest with success", valid, args{resultTestBytes}, false},
+		{"unmarshal invalid AddDeviceServiceRequest, empty data", UpdateDeviceServiceRequest{}, args{[]byte{}}, true},
+		{"unmarshal invalid AddDeviceServiceRequest, string data", UpdateDeviceServiceRequest{}, args{[]byte("Invalid AddDeviceServiceRequest")}, true},
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceservice_test.go:156]: Chnage `AddDeviceServiceRequest ` to `UpdateDeviceServiceRequest`

File: v2/dtos/requests/deviceservice_test.go [hunk 9]
```diff
@@ -115,3 +138,122 @@ func Test_AddDeviceServiceReqToDeviceServiceModels(t *testing.T) {
...
+	fmt.Println(string(resultTestBytes))
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var expected = tt.addDeviceService
+			err := tt.addDeviceService.UnmarshalJSON(tt.args.data)
+			if tt.wantErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, expected, tt.addDeviceService, "Unmarshal did not result in expected AddDeviceServiceRequest.", err)
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceservice_test.go:167]: Chnage `AddDeviceServiceRequest ` to `UpdateDeviceServiceRequest`

File: v2/dtos/requests/deviceservice_test.go [hunk 10]
```diff
@@ -115,3 +138,122 @@ func Test_AddDeviceServiceReqToDeviceServiceModels(t *testing.T) {
...
+		DeviceService UpdateDeviceServiceRequest
+		expectError   bool
+	}{
+		{"valid AddDeviceServiceRequest", valid, false},
+		{"valid AddDeviceServiceRequest without Id", validWithoutId, false},
+		{"valid AddDeviceServiceRequest without name", validWithoutName, false},
+		{"invalid AddDeviceServiceRequest, no Request Id", noReID, true},
+		{"invalid AddDeviceServiceRequest, no Id and Name", noIdAndName, true},
+		{"invalid AddDeviceServiceRequest, invalid OperatingState", invalidOperatingState, true},
+		{"invalid AddDeviceServiceRequest, invalid AdminState", invalidAdminState, true},
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceservice_test.go:200]: Chnage `AddDeviceServiceRequest ` to `UpdateDeviceServiceRequest`

File: v2/dtos/requests/deviceservice_test.go [hunk 11]
```diff
@@ -115,3 +138,122 @@ func Test_AddDeviceServiceReqToDeviceServiceModels(t *testing.T) {
...
+        "requestId":"7a1707f0-166f-4c4b-bc9d-1d54c74e0137",
+		"service":{
+			"name":"test device service"
+		}
+	}`
+	var req UpdateDeviceServiceRequest
+
+	err := req.UnmarshalJSON([]byte(reqJson))
+
+	assert.Equal(t, nil, err)
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceservice_test.go:221]: Use `require.NoError(...)`

File: v2/dtos/requests/deviceservice_test.go [hunk 12]
```diff
@@ -115,3 +138,122 @@ func Test_AddDeviceServiceReqToDeviceServiceModels(t *testing.T) {
...
+		"service":{
+			"name":"test device",
+			"labels":[]
+		}
+	}`
+	var req UpdateDeviceServiceRequest
+
+	err := req.UnmarshalJSON([]byte(reqJson))
+
+	assert.Equal(t, nil, err)
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceservice_test.go:241]: Use `require.NoError(...)`

File: v2/dtos/requests/deviceservice_test.go [hunk 13]
```diff
@@ -115,3 +138,122 @@ func Test_AddDeviceServiceReqToDeviceServiceModels(t *testing.T) {
...
+	// Nil field checking is used to update with patch
+	assert.True(t, req.Service.BaseAddress == nil)
+	assert.True(t, req.Service.AdminState == nil)
+	assert.True(t, req.Service.OperatingState == nil)
+	assert.True(t, req.Service.Labels == nil)
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceservice_test.go:226]: use `assert.Nil(...)`

File: v2/dtos/requests/deviceservice_test.go [hunk 14]
```diff
@@ -115,3 +138,122 @@ func Test_AddDeviceServiceReqToDeviceServiceModels(t *testing.T) {
...
+			"labels":[]
+		}
+	}`
+	var req UpdateDeviceServiceRequest
+
+	err := req.UnmarshalJSON([]byte(reqJson))
+
+	assert.Equal(t, nil, err)
+	// Empty slice is used to remove the data
+	assert.True(t, req.Service.Labels != nil)
```
Comment (lenny-intel) [inline v2/dtos/requests/deviceservice_test.go:243]: use `assert.NotNil(...)`

Comment (weichou1229) [reply v2/dtos/requests/deviceservice_test.go]: Finished all request changes. 
Thanks for catching lots of typo and misused.

**Pull request merged**